### PR TITLE
Custom methoddescription

### PR DESCRIPTION
**Pull request by jfy133**
## PR checklist

This adds another custom module, this time attempting to make a semi-automated 'methods description' that is displayed in MultiQC, with which a user could copy over the text into their own publication. 

We can thus use this as a method for ensuring 'best practise' descriptions on a pipeline was run, to ensure maximal reproducibility.

This is a WIP, and suggestions/changes are welcome.  

Example output:

image

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/custom/methodsdescription/main.nf [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    section_name: 'Methods Description'
+    -->
+    <p>
+    Data was processed using the nf-core (Ewels et al. 2020) pipeline ${workflow.manifest.name} v${workflow.manifest.version}$doi.
+
+    The pipeline was executed with Nextflow (v${nextflow.version}; Di Tommaso et al. 2017) with the following command:</p>
+
+    <pre><code>
+    ${workflow.commandLine}
+    </code></pre>
```
Comment (ewels) [inline modules/custom/methodsdescription/main.nf:27]: Nice! I didn't know about this one..

```suggestion
    <pre><code>${workflow.commandLine}</code></pre>
```

File: modules/custom/methodsdescription/main.nf [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    """
+    cat <<-METHOD_DESCRIPTION > method_description_mqc.html
+    <!--
+    id: 'methods-description'
+    section_name: 'Methods Description'
+    -->
+    <p>
+    Data was processed using the nf-core (Ewels et al. 2020) pipeline ${workflow.manifest.name} v${workflow.manifest.version}$doi.
+
+    The pipeline was executed with Nextflow (v${nextflow.version}; Di Tommaso et al. 2017) with the following command:</p>
```
Comment (ewels) [inline modules/custom/methodsdescription/main.nf:23]: Not sure that we need the references here as well as below?
```suggestion
    Data was processed using ${workflow.manifest.name} v${workflow.manifest.version}$doi.

    The pipeline was executed with Nextflow v${nextflow.version} with the following command:</p>
```

File: modules/custom/methodsdescription/main.nf [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    </ul>
+
+    <blockquote>
+    <b>Notes:</b><br>
+    <ul>
+        <li>If present, make sure to replace the pipeline DOI with correct reference information.</li>
+        <li>The command above does not include parameters contained in any configs that may have been used. Ensure the config file is also uploaded with your publication.</li>
+        <li>You should also cite all software used within this run. Check the 'Software Versions' of this report to get version information.</li>
+    </ul>
+    </blockquote>
```
Comment (ewels) [inline modules/custom/methodsdescription/main.nf:42]:
```suggestion
    <div class="alert alert-info">
    <ul>
        <li>If present, make sure to replace the pipeline DOI with correct reference information.</li>
        <li>The command above does not include parameters contained in any config files that may have been used. Please ensure that these are included in your publication.</li>
        <li>Please also cite the software tools used within this run. These are also listed in this MultiQC report.</li>
    </ul>
    </div>
```

File: modules/custom/methodsdescription/main.nf [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    <b>References</b>
+    <ul>
+        <li>Di Tommaso, P., Chatzou, M., Floden, E. W., Barja, P. P., Palumbo, E., & Notredame, C. (2017). Nextflow enables reproducible computational workflows. Nature Biotechnology, 35(4), 316–319. https://doi.org/10.1038/nbt.3820 </li>
+        <li>Ewels, P. A., Peltzer, A., Fillinger, S., Patel, H., Alneberg, J., Wilm, A., Garcia, M. U., Di Tommaso, P., & Nahnsen, S. (2020). The nf-core framework for community-curated bioinformatics pipelines. Nature Biotechnology, 38(3), 276–278. https://doi.org/10.1038/s41587-020-0439-x </li>
+    </ul>
+
+    <blockquote>
+    <b>Notes:</b><br>
+    <ul>
+        <li>If present, make sure to replace the pipeline DOI with correct reference information.</li>
```
Comment (ewels) [inline modules/custom/methodsdescription/main.nf:38]: We know whether it's present or not, right? Can we make this line conditional as we do in the text above with `$doi`?

File: modules/custom/methodsdescription/main.nf [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    path "versions.yml" , emit: versions
+
+    script:
+    def args = task.ext.args ?: ''
+    def doi = workflow.manifest.doi ? " (${workflow.manifest.doi})" : ""
+
+    """
+    cat <<-METHOD_DESCRIPTION > method_description_mqc.html
+    <!--
+    id: 'methods-description'
```
Comment (ewels) [inline modules/custom/methodsdescription/main.nf:17]: To group together with other MultiQC sections generated by the pipeline. See https://github.com/nf-core/tools/pull/1389/

```suggestion
    parent_id: '${workflow.manifest.name.replace('/','-')}_run_details
    id: 'methods-description'
```

**Review by ewels — COMMENTED** [review]
Awesome! Again, as with https://github.com/nf-core/tools/pull/1389/ I feel like this probably is better suited to live in a `lib/` function call rather than in a process, as it's very simple. As a bonus, that means all pipelines will get it automatically. 

It should be pretty trivial to port there, you've done the hard work already.

In fact, it could just be part of the _Workflow summary_ section perhaps? That's already [printing custom HTML](https://github.com/nf-core/tools/blob/e563811adb292b07ec933a52611e937ce9ef5481/nf_core/pipeline-template/lib/WorkflowPipeline.groovy#L19-L44), so it could just be lifted directly into it.

**Comment by Midnighter** [comment]
I don't have any comments on the code part but I think it'd be neat if the generated HTML was a template that could easily be replaced on a per-pipeline basis. I think a methods description should include arguments used but that has to be decided for each pipeline individually and cannot be solved generally.

**Comment by jfy133** [comment]
I completely forgot about this, and now understand what is requested. I see what @Midnighter means about wanting to tweak on a per-pipeline basis/have an entirely custom text for a pipeline, but I'll start with just coming up with a basic default if I can.

**Comment by jfy133** [comment]
Ok, `lib/` version incoming! Proof of concept: https://github.com/jfy133/nf-core-testpipeline/tree/methods-summary

Self note:

- Need conditional DOI / pipeline citation insertion
- Need to work out if can have custom HTML message to ensure two space indent at beginning of lines, and evalute variables with the HTML (`.text` doesn't allow this)

```
nextflow run ../main.nf -profile singularity,test --outdir ./results --custom_methods_description custom_methods.html 
```

**Comment by jfy133** [comment]
Replacing with: https://github.com/nf-core/tools/pull/1749

**Pull request closed without merging**